test/WorkflowCore.UnitTests/BasePersistenceFixture.cs

```diff
@@ -92,5 +93,58 @@ public void PersistWorkflow()
             var current = Subject.GetWorkflowInstance(workflowId).Result;
             current.ShouldBeEquivalentTo(newWorkflow);
         }
+
+        [Fact]
+        public void ConcurrentPersistWorkflow()
+        {
+            var subject = Subject; // Don't initialize in the thread.
+
+            var tasks = new List<Task<string>>();
+
+            for (int i = 0; i < 30; i++)
+            {
+                tasks.Add(Task.Run(() =>
+                {
+                    try
+                    {
+                        var oldWorkflow = new WorkflowInstance()
+                        {
+                            Data = new { Value1 = 7 },
+                            Description = "My Description",
+                            Status = WorkflowStatus.Runnable,
+                            NextExecution = 0,
+                            Version = 1,
+                            WorkflowDefinitionId = "My Workflow",
+                            CreateTime = new DateTime(2000, 1, 1).ToUniversalTime()
+                        };
+                        oldWorkflow.ExecutionPointers.Add(new ExecutionPointer()
+                        {
+                            Id = Guid.NewGuid().ToString(),
+                            Active = true,
+                            StepId = 0
+                        });
+                        var workflowId = subject.CreateNewWorkflow(oldWorkflow).Result;
+                        var newWorkflow = Utils.DeepCopy(oldWorkflow);
+                        newWorkflow.NextExecution = 7;
+                        newWorkflow.ExecutionPointers.Add(new ExecutionPointer() { Id = Guid.NewGuid().ToString(), Active = true, StepId = 1 });
+
+                        subject.PersistWorkflow(newWorkflow).Wait(); // It will throw an exception if the persistence provider occurred resource competition.
+
+                        return workflowId;
+                    }
+                    catch (Exception)
+                    {
+                        return null;
+                    }
+                }));
+            }
+
+            Task.WhenAll(tasks).Wait();
+
+            foreach (var task in tasks)
+            {
+                task.Result.Should().NotBeNull();
+            }
```

Comment:
Awesome... instead of catching the exception and returning null... could we perhaps let the exception bubble up and assert it did not happen with fluent assertions?
http://fluentassertions.com/documentation.html#exceptions